### PR TITLE
allow operator to manage permissions based on namespace labels

### LINKED ISSUE #352: Automating K8s RBAC for Argo CD instances
**Is your feature request related to a problem? Please describe.**

Proposed by @jannfis 

Currently, the integration pattern for the Operator is the following

* Use the out-of-the-box (OOTB) Argo CD instance to manage some cluster resources
* Use a dedicated Argo CD instance to manage an application in the instance’s target namespace

When a user creates a new Argo CD instance with a target namespace of e.g. “myapp”, the operator will automatically create Kubernetes RBAC permissions in the target namespace in the form of a Role and a RoleBinding for the ServiceAccount used by the Argo CD instance’s application controller. 

This is fine if the target namespace is the only namespace being managed by the Argo CD instance. However, users also want to use this instance to manage multiple namespaces in the cluster. The current recommendation is to create the Role and RoleBinding objects manually. This can be cumbersome to set up, complicated to manage and should be automated.

**Describe the solution you'd like**

### Overview

We should give users an easy way to configure namespaces that should be managed by a given Argo CD instance.

I envision this by using a label on any namespace:
`argocd.argoproj.io/managed-by: <name of Argo CD instance>`

Each namespace should be only manageable by a single Argo CD instance, but one Argo CD instance will manage N namespaces, effectively establishing a 1:n relationship.

This pattern will work best with people creating new projects (namespaces) manually on their cluster. They can simply give a specific Argo CD instance named “example” permissions to the namespace “myproj” by issuing the following command:

`$ kubectl label namespace myproj argocd.argoproj.io/managed-by=example`

Likewise, revoking permissions is as simple as running

`$ kubectl label namespace myproj argocd.argoproj.io/managed-by-`

By leveraging project templates, this could be already automated at project creation time.

Given a scenario where the Argo CD instance has permissions to create Namespace resources in the cluster, and such a Namespace resource exists in the application’s manifests, it is not clear yet whether the approach will work. Argo CD will create the namespace from the resource, possibly with the `managed-by` annotation, as the first action of the initial sync and continue with the other resources. However, the Operator might not have created appropriate RBAC resources yet, and the sync will fail due to permission denied. I believe the auto-retry feature on sync will succeed the sync on the next retry, however, this has to be proven.

### RBAC resources managed by the operator

The Role and RoleBinding resources created by the Operator should be clearly indicated as such. The Operator must explicitly set the ownerReferences on the resources, and the proposal is to use an additional label for easier indexing and retrieval of the resources. The proposed label is the same as for the Namespace resource:

    `argocd.argoproj.io/managed-by: <name of Argo CD instance>`

The `ownerReferences` will instruct the Garbage Collector of Kubernetes to remove the resources whenever the Operator’s Operand is uninstalled from the cluster, so that a proper cleanup is done automatically.

In the MVP implementation of this proposal, the Role should give namespace admin permissions. Going forward, we can think of ways to provide templates for Roles that will get reconciled, so that permissions can be further locked down.
Reconciliation process
The  Operator will perform reconciliation of appropriate Role and RoleBinding namespaces that are annotated as proposed above.

The Operator will establish a watch on Namespace resources, and if it sees such a label is being added, modified or removed, reconcile appropriate Cluster and ClusterRole resources into that namespace.

The Operator knows which instances of Argo CD currently exist, and it also has the appropriate permissions for reconciling Cluster and ClusterRole resources everywhere.

When a namespace is annotated with the `argocd.argoproj.io/managed-by: <value> `label, the  Operator will:

- Check for existing Role and RoleBinding resources in the namespace that carry a label of `argocd.argoproj.io/managed-by`, and if it matches the value of the annotation with the same name set on the Namespace

- If values differ between the `argocd.argoproj.io/managed-by` on the RBAC resources and the Namespace resource, the operator will delete the existing Role and RoleBinding resources from the Namespace

- If values are the same, the Operator will reconcile the existing Role and RoleBinding resources to the version as deemed correct by the Operator

- If no such Role and RoleBinding with a label `argocd.argoproj.io/managed-by` are found, the Operator will create them

### Further possible enhancements

Permission levels / templates
Another annotation, e.g. `argocd.argoproj.io/permission-level`: <value>, could be specified on the namespace to indicate what kind of permissions Argo CD application controller should receive for this namespace.

Using this annotation, we could also provide some predefined permission levels additional to “admin” permissions, and also let the user define their own permission templates in the operand, which they can refer to from the annotation.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by shubhamagarwal19**
**What type of PR is this?**
> /kind enhancement

**What does this PR do / why we need it**:
This PR addresses the need to manually provide the permissions to ArgoCD instance for managing a new namespace. This feature would allow the user to simply label a namespace with the label `argocd.argoproj.io/managed-by: <namespace of argocd>` and the operator would create the required permissions. 

**Have you updated the necessary documentation?**

* [x] Documentation update is required by this PR.
* [ ] Documentation has been updated.

**Which issue(s) this PR fixes**:

Fixes  #352

**How to test changes / Special notes to the reviewer**:
// TODO

- [x] Add Unit Tests

### CODE REVIEW

File: pkg/controller/argocd/role.go [hunk 1]
```diff
@@ -55,19 +57,19 @@ func newClusterRole(name string, rules []v1.PolicyRule, cr *argoprojv1a1.ArgoCD)
 
 // reconcileRoles will ensure that all ArgoCD Service Accounts are configured.
 func (r *ReconcileArgoCD) reconcileRoles(cr *argoprojv1a1.ArgoCD) (role *v1.Role, err error) {
```
Comment (wtam2018) [inline pkg/controller/argocd/role.go:59]: This function probably should not return `role`.

File: pkg/controller/argocd/role.go [hunk 2]
```diff
@@ -83,30 +85,54 @@ func (r *ReconcileArgoCD) reconcileRoles(cr *argoprojv1a1.ArgoCD) (role *v1.Role
...
 
-	if name == dexServer && isDexDisabled() {
-		// Delete any existing Role created for Dex
-		return role, r.client.Delete(context.TODO(), role)
+		if name == dexServer && isDexDisabled() {
+			// Delete any existing Role created for Dex
+			if err := r.client.Delete(context.TODO(), role); err != nil {
+				return nil, err
+			}
+			roles = append(roles, role)
```
Comment (wtam2018) [inline pkg/controller/argocd/role.go:126]: Doing an `append` here does not look right.

File: pkg/controller/argocd/secret.go [hunk 3]
```diff
@@ -423,12 +440,22 @@ func (r *ReconcileArgoCD) reconcileClusterPermissionsSecret(cr *argoprojv1a1.Arg
...
 			} else {
-				return nil
+				ns := strings.Split(string(s.Data["namespaces"]), ",")
+				for _, n := range namespaces {
+					if !containsString(ns, strings.TrimSpace(n)) {
+						ns = append(ns, strings.TrimSpace(n))
+					}
+					sort.Strings(ns)
+					s.Data["namespaces"] = []byte(strings.Join(ns, ","))
+					return r.client.Update(context.TODO(), &s)
```
Comment (wtam2018) [inline pkg/controller/argocd/secret.go:457]: I don't think we want to return yet.  Also, we should only need to call `Update` once outside of the for-loop.

**Review by wtam2018 — CHANGES_REQUESTED** [review]
I think you forgot about writing unit tests.

File: pkg/controller/argocd/role.go (hunk 1 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/role.go:59]: These roles are needed when we reconcile `RoleBindings`.

File: pkg/controller/argocd/secret.go (hunk 3 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/secret.go:457]: Technically, there would always be only one instance of the Secret for the default cluster server, but if there are multiple instances (which is highly unlikely) we would update the first available instance. 
Do you think we should be trying to update all the secrets if there are multiple instances?

File: pkg/controller/argocd/custommapper.go [hunk 4]
```diff
@@ -96,3 +97,30 @@ func (r *ReconcileArgoCD) tlsSecretMapper(o handler.MapObject) []reconcile.Reque
 
 	return result
 }
+
+func (r *ReconcileArgoCD) namespaceResourceMapper(o handler.MapObject) []reconcile.Request {
+	var result = []reconcile.Request{}
+
+	labels := o.Meta.GetLabels()
+	if v, ok := labels[common.ArgoCDManagedNamespaceLabel]; ok {
+		argocds := &argoprojv1alpha1.ArgoCDList{}
```
[the remark this answers is not in the document]
Comment (shubhamagarwal19) [reply pkg/controller/argocd/custommapper.go:106]: According to the design it is was decided to use the <source namespace> as the value for the label to identify the ArgoCD instance. 
But it does seem like a better idea to use an identifier `namespace/name`. That way i wouldn't need to list ArgoCD instances, i can directly query for the particular instance. @jannfis @sbose78 thoughts?

[the remark this answers is not in the document]
Comment (jannfis) [reply pkg/controller/argocd/custommapper.go:106]: That sounds tempting, but we need to take care not to hit the 63 characters limit for labels here. Names of namespaces can become pretty long already (they max out at 63 chars as well). So if someone has a long namespace name, this may not work.

File: pkg/controller/argocd/argocd_controller.go [hunk 5]
```diff
@@ -94,6 +96,21 @@ func (r *ReconcileArgoCD) Reconcile(request reconcile.Request) (reconcile.Result
 		return reconcile.Result{}, err
 	}
 
+	namespace := &corev1.Namespace{}
```
[the remark this answers is not in the document]
Comment (jannfis) [reply pkg/controller/argocd/argocd_controller.go:99]: I think it's up to the user to remove the label, because they put it there initially?

[the remark this answers is not in the document]
Comment (jannfis) [reply pkg/controller/argocd/argocd_controller.go:99]: Err. Oops. OK, I see now :smile: We put the label on the namespace where Argo CD is installed to. I agree that upon uninstallation of the Operand, that one label should get removed as well. Maybe in the finalizer logic for `ArgoCD` resource?

File: pkg/controller/argocd/custommapper.go [hunk 6]
```diff
@@ -96,3 +97,30 @@ func (r *ReconcileArgoCD) tlsSecretMapper(o handler.MapObject) []reconcile.Reque
 
 	return result
 }
+
+func (r *ReconcileArgoCD) namespaceResourceMapper(o handler.MapObject) []reconcile.Request {
```
Comment (jannfis) [inline pkg/controller/argocd/custommapper.go:101]: Can you please put some comments on this function, describing what it does and why?

File: pkg/controller/argocd/role.go (hunk 1 above)
Comment (jannfis) [reply pkg/controller/argocd/role.go:59]: I am confused about this method as well. It basically never returns any role, if I interpret the code correctly.

if `r.reconcileRole()` returns `err != nil`, `role` would be `nil` as well, right? So basically, all it ever returns is `nil` for `role` return value. :thinking:

File: pkg/controller/argocd/role.go [hunk 7]
```diff
@@ -83,30 +85,54 @@ func (r *ReconcileArgoCD) reconcileRoles(cr *argoprojv1a1.ArgoCD) (role *v1.Role
 }
 
 // reconcileRole
-func (r *ReconcileArgoCD) reconcileRole(name string, policyRules []v1.PolicyRule, cr *argoprojv1a1.ArgoCD) (*v1.Role, error) {
-	role := newRole(name, policyRules, cr)
-	if err := applyReconcilerHook(cr, role, ""); err != nil {
+func (r *ReconcileArgoCD) reconcileRole(name string, policyRules []v1.PolicyRule, cr *argoprojv1a1.ArgoCD) ([]*v1.Role, error) {
```
Comment (jannfis) [inline pkg/controller/argocd/role.go:88]: I think we really need some documentation within this method. Why is any role created, or not. This is a critical path function. Please help others who need to modify this in the future to understand _why_ something is done the way it is done. Don't assume any context knowledge.

File: pkg/controller/argocd/rolebinding.go [hunk 8]
```diff
@@ -83,68 +83,79 @@ func (r *ReconcileArgoCD) reconcileRoleBindings(cr *argoprojv1a1.ArgoCD) error {
 }
 
 func (r *ReconcileArgoCD) reconcileRoleBinding(name string, rules []v1.PolicyRule, cr *argoprojv1a1.ArgoCD) error {
```
Comment (jannfis) [inline pkg/controller/argocd/rolebinding.go:85]: Same here as with `reconcileRole()`. Please document it. Help others understanding it.

File: pkg/controller/argocd/secret.go [hunk 9]
```diff
@@ -423,12 +440,22 @@ func (r *ReconcileArgoCD) reconcileClusterPermissionsSecret(cr *argoprojv1a1.Arg
...
 	for _, s := range clusterSecrets.Items {
 		// check if cluster secret with default server address exists
-		// do nothing if exists.
+		// update the list of namespaces if value differs.
 		if string(s.Data["server"]) == common.ArgoCDDefaultServer {
 			if clusterConfigInstance {
-				r.client.Delete(context.TODO(), &s)
+				if err := r.client.Delete(context.TODO(), &s); err != nil {
+					return err
+				}
```
Comment (jannfis) [inline pkg/controller/argocd/secret.go:448]: I was just wondering whether we want to only delete `namespaces` key from the secret here, instead of completely deleting the secret. User may have created secret to change other properties, such as cluster shard index. But let's make this another PR.

File: pkg/controller/argocd/secret.go (hunk 3 above)
Comment (jannfis) [reply pkg/controller/argocd/secret.go:457]: We do assume (and support) only one ArgoCD instance per namespace, so I think it's ok.

**Review by jannfis — CHANGES_REQUESTED** [review]
Thanks @shubhamagarwal19 

I'm addressing some minor issues below. PTAL.

File: pkg/controller/argocd/role.go (hunk 7 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/role.go:88]: `reconcileRole` returns a list of roles for all namespaces which is managed by a single ArgoCD instance. So for example if `foo and bar` is managed by an instance of ArgoCD, then it would return a list of roles for `application-controller/argocd-server ... etc` for each namespace. 
This list of roles is then used by `reconcileRolebinding` to create roleBindings for each role. 

Although i see this as an issue, this would create issues while trying to remove the rolebindings say for `DEX` if it gets disabled.

File: pkg/controller/argocd/secret.go (hunk 9 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/secret.go:448]: Could there be any other configuration in the secret that could affect the functioning of a `ClusterConfig ArgoCD instance` ?

File: pkg/controller/argocd/custommapper.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sbose78) [reply pkg/controller/argocd/custommapper.go:106]: That one extra API Call should be fine :) the lesser the user does, the more accurate he will be.

File: pkg/common/keys.go [hunk 10]
```diff
@@ -200,4 +200,7 @@ const (
 
 	// ArgoCDSecretTypeLabel is needed for cluster secrets
 	ArgoCDSecretTypeLabel = "argocd.argoproj.io/secret-type"
+
+	// ArgoCDManagedNamespaceLabel is needed to identify namespace managed by an instance on ArgoCD
+	ArgoCDManagedNamespaceLabel = "argocd.argoproj.io/managed-by"
```
[the remark this answers is not in the document]
Comment (shubhamagarwal19) [reply pkg/common/keys.go]: done! thanks

File: pkg/controller/argocd/argocd_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (shubhamagarwal19) [reply pkg/controller/argocd/argocd_controller.go]: Thanks, @wtam2018 @jannfis!! As suggested, added the code to delete the label from the namespace in the `finalizer logic` !

File: pkg/controller/argocd/role.go (hunk 2 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/role.go]: removed it!! Thanks, I had added this for something that I had in my mind, but that wouldn't work!!

File: pkg/controller/argocd/custommapper.go (hunk 6 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/custommapper.go:103]: added the comments! Thanks

File: pkg/controller/argocd/rolebinding.go (hunk 8 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/rolebinding.go:87]: added comments!

File: pkg/controller/argocd/role.go (hunk 1 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/role.go:59]: We actually don't need the list of `roles` when we are calling the `reconcileRole` explicitly and just need to make sure that the role was created without any errors!! 
But when we call `reconcileRole` from within the `reconcileRoleBindings`, we need the list of roles that were created which would eventually be used to create the `RoleBindings`!

This would make more sense when we refactor the ArgoCD Operator code, which wouldn't reconcile the `Role/RoleBinding/ServiceAccount` individually, rather, it would call a function something like `reconcileArgoCDPermissions` and within the function, it would reconcile all the resources in the order `ServiceAccount -> Roles -> RoleBindings`

File: pkg/controller/argocd/role.go (hunk 7 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/role.go:89]: Added comments!

**Comment by shubhamagarwal19** [comment]
> I think you forgot about writing unit tests.

Done!! added tests!
